File: src/minimumGL.nim
import winlean4k, openGL4k2

const
  ScreenWidth   = 640
  ScreenHeight  = 480

let
  pfd           = PIXELFORMATDESCRIPTOR(
    nSize:          sizeof(PIXELFORMATDESCRIPTOR).uint16,
    nVersion:       1'u16,
    dwFlags:        PFD_DRAW_TO_WINDOW or
                    PFD_SUPPORT_OPENGL or
                    PFD_DOUBLEBUFFER,
    iPixelType:     PFD_TYPE_RGBA,
    cColorBits:     24,
    cRedBits:       0, cRedShift:   0,
    cGreenBits:     0, cGreenShift: 0,
    cBlueBits:      0, cBlueShift:  0,
    cAlphaBits:     0, cAlphaShift: 0,
    cAccumBits:     0,
    cAccumRedBits:  0,
    cAccumGreenBits:0,
    cAccumBlueBits: 0,
    cAccumAlphaBits:0,
    cDepthBits:     24,
    cStencilBits:   0,
    cAuxBuffers:    0,
    iLayerType:     PFD_MAIN_PLANE,
    bReserved:      0,
    dwLayerMask:    0,
    dwVisibleMask:  0,
    dwDamageMask:   0)

proc initScreen() {.inline.} =
  let hWnd = CreateWindowA(
    "STATIC".cstring, nil,
    WS_POPUP or WS_VISIBLE, 0, 0,
    ScreenWidth, ScreenHeight, nil, nil, nil, nil)
  let hdc = GetDC(hWnd)

  discard SetPixelFormat(hdc, ChoosePixelFormat(hdc, unsafeAddr pfd), unsafeAddr pfd)
  discard wglMakeCurrent(hdc, wglCreateContext(hdc))

proc WinMainCRTStartup() {.exportc.} =
  initScreen()
  let version = glGetString(GL_VERSION)
  discard MessageBoxA(nil, cast[cstring](version), cstring"minimumGL", 0)
  # Process keep alive if ExitProcess API was not called.
  ExitProcess(0)

when not defined(danger):
  WinMainCRTStartup()

File: csrc/minimumGL.c
/*
 * Minimum C program.
 * Nim program can be as small as this program?
 * Run minimum_build.bat on VS2015 x86 Native Tools command prompt to build this code.
 */

#include <windows.h>
#include <GL/gl.h>
#include <GL/wglext.h>

#define WIDTH 640
#define HEIGHT 480

static PIXELFORMATDESCRIPTOR pfd =
{
	sizeof(PIXELFORMATDESCRIPTOR),
	1,									          //OpenGL version
	PFD_DRAW_TO_WINDOW |
	PFD_SUPPORT_OPENGL |
	PFD_DOUBLEBUFFER,
	PFD_TYPE_RGBA,
	24,									//Color bits
	0, 0,
	0, 0,
	0, 0,
	0, 0,
	0,
	0, 0, 0, 0,
	24,									//Z buffer
	0,
	0,
	PFD_MAIN_PLANE,
	0,
	0, 0, 0
};

void initScreen() {
	HWND hWnd =	CreateWindowA("STATIC", 0,
		WS_POPUP|WS_VISIBLE, 0, 0,
		WIDTH, HEIGHT, NULL, NULL, NULL, NULL);

	HDC hdc = GetDC(hWnd);

	SetPixelFormat(hdc, ChoosePixelFormat(hdc, &pfd), &pfd);

	wglMakeCurrent(hdc, wglCreateContext(hdc));
}

void WinMainCRTStartup(void) {
  initScreen();
	MessageBoxA(0, glGetString(GL_VERSION), "minimumGL", 0);
  //Process keep alive if ExitProcess API was not called.
  ExitProcess(0);
}
